File: src/systems/paths/LinearPath.ts
import { PathData, PathType } from "./PathData";

/**
 * Straight line movement from start to end point
 * Exact port of C# LinearPath implementation
 */
export class LinearPath {
  private readonly start: [number, number];
  private readonly end: [number, number];
  private readonly fishId: number;
  private readonly seed: number;
  private readonly startTick: number;
  private readonly speed: number;

  constructor(
    fishId: number,
    seed: number,
    startTick: number,
    speed: number,
    start: [number, number],
    end: [number, number],
  ) {
    this.fishId = fishId;
    this.seed = seed;
    this.startTick = startTick;
    this.speed = speed;
    this.start = start;
    this.end = end;
  }

  /**
   * Get position along the linear path at normalized time t
   * Matches C# LinearPath.GetPosition (lines 27-32)
   */
  getPosition(t: number): [number, number] {
    const x = this.start[0] + (this.end[0] - this.start[0]) * t;
    const y = this.start[1] + (this.end[1] - this.start[1]) * t;
    return [x, y];
  }

  getPathType(): PathType {
    return PathType.Linear;
  }

  /**
   * Get serializable path data
   * Matches C# LinearPath.GetPathData (lines 37-55)
   */
  getPathData(): PathData {
    const distance = Math.sqrt(
      Math.pow(this.end[0] - this.start[0], 2) +
        Math.pow(this.end[1] - this.start[1], 2),
    );

    return {
      fishId: this.fishId,
      pathType: PathType.Linear,
      seed: this.seed,
      startTick: this.startTick,
      speed: this.speed,
      controlPoints: [this.start, this.end],
      duration: distance / this.speed,
      loop: false,
    };
  }
}

File: Server/Systems/Paths/LinearPath.cs
using OceanKing.Server.Models;

namespace OceanKing.Server.Systems.Paths;

/// <summary>
/// Straight line movement from start to end point
/// </summary>
public class LinearPath : IPath
{
    private readonly float[] _start;
    private readonly float[] _end;
    private readonly int _fishId;
    private readonly int _seed;
    private readonly int _startTick;
    private readonly float _speed;
    
    public LinearPath(int fishId, int seed, int startTick, float speed, float[] start, float[] end)
    {
        _fishId = fishId;
        _seed = seed;
        _startTick = startTick;
        _speed = speed;
        _start = start;
        _end = end;
    }
    
    public float[] GetPosition(float t)
    {
        // Linear interpolation between start and end
        float x = _start[0] + (_end[0] - _start[0]) * t;
        float y = _start[1] + (_end[1] - _start[1]) * t;
        return new[] { x, y };
    }
    
    public PathType GetPathType() => PathType.Linear;
    
    public PathData GetPathData()
    {
        float distance = MathF.Sqrt(
            MathF.Pow(_end[0] - _start[0], 2) + 
            MathF.Pow(_end[1] - _start[1], 2)
        );
        
        return new PathData
        {
            FishId = _fishId,
            PathType = PathType.Linear,
            Seed = _seed,
            StartTick = _startTick,
            Speed = _speed,
            ControlPoints = new[] { _start, _end },
            Duration = distance / _speed,
            Loop = false
        };
    }
}
